src/kernels/common/baseJupyterSession.ts

```diff
@@ -83,28 +91,30 @@ export abstract class BaseJupyterSession implements IJupyterSession {
     protected get session(): ISessionWithSocket | undefined {
         return this._session;
     }
+    public get kernelId(): string {
+        return this.session?.kernel?.id || '';
+    }
     public get kernel(): Kernel.IKernelConnection | undefined {
-        return this._session?.kernel || undefined;
+        if (this._wrappedKernel) {
+            return;
+        }
+        if (!this._session?.kernel) {
+            return;
+        }
+        this._wrappedKernel = new KernelConnectionWrapper(this._session.kernel, this.disposables);
+        return this._wrappedKernel;
     }
+
     public get kernelSocket(): Observable<KernelSocketInformation | undefined> {
         return this._kernelSocket;
     }
     public get onSessionStatusChanged(): Event<KernelMessage.Status> {
         return this.onStatusChangedEvent.event;
     }
-    public get onIOPubMessage(): Event<KernelMessage.IIOPubMessage> {
```

Comment:
We need a way to hook up to the `iopub` and `anymessage` events in the underlyiing Jupyter kernel connection.
This property `onIOPubMessage` is a custom event we wire up. If I were to introduce `anyMessage` event then i'd have to handle kernel restarts all over again, tomorrow if we wanted to handle other events, then all over gain etc... 

Also, as we expose the underlying jupyter kernel (Kernel.IKenrelConnection) everywhere it makes sense to just hook to those events.

Hence created a kernel wrapper so that we have a single instance, of the Kernel.KernleConnection on which we can monitor the events (basically moved onIOPubMessage & similar events into the kernel.IKernelConnection and have that keep track of all the event handlers even when kernels restart)